src/lib/vasi-sync/src/scchannel.rs

```diff
@@ -0,0 +1,254 @@
+        const WRITING: u8 = ChannelContentsState::Writing as u8;
+        const READY: u8 = ChannelContentsState::Ready as u8;
+        const READING: u8 = ChannelContentsState::Reading as u8;
+        match value {
+            EMPTY => ChannelContentsState::Empty,
+            WRITING => ChannelContentsState::Writing,
+            READY => ChannelContentsState::Ready,
+            READING => ChannelContentsState::Reading,
+            _ => panic!("Bad value {value}"),
+        }
+    }
+}
+
+// bit flags in ChannelState
+const WRITER_DIED: u32 = 0x1 << 9;
+const HAS_SLEEPER: u32 = 0x1 << 10;
+
+#[derive(Debug, Eq, PartialEq, Copy, Clone)]
+struct ChannelState {
+    writer_died: bool,
+    has_sleeper: bool,
+    contents_state: ChannelContentsState,
+}
+
+impl From<u32> for ChannelState {
+    fn from(value: u32) -> Self {
+        let has_sleeper = (value & HAS_SLEEPER) != 0;
+        let writer_died = (value & WRITER_DIED) != 0;
+        let contents_state = ((value & 0xff) as u8).into();
+        ChannelState {
+            has_sleeper,
+            writer_died,
+            contents_state,
+        }
+    }
+}
+
+impl From<ChannelState> for u32 {
+    fn from(value: ChannelState) -> Self {
+        let has_sleeper = if value.has_sleeper { HAS_SLEEPER } else { 0 };
+        let writer_died = if value.writer_died { WRITER_DIED } else { 0 };
+        writer_died | has_sleeper | (value.contents_state as u32)
+    }
+}
+
+struct AtomicChannelState(AtomicU32);
+impl AtomicChannelState {
+    pub fn new() -> Self {
+        Self(AtomicU32::new(
+            ChannelState {
+                has_sleeper: false,
+                writer_died: false,
+                contents_state: ChannelContentsState::Empty,
+            }
+            .into(),
+        ))
+    }
+
+    fn transition_contents_state(
+        &self,
+        from: ChannelContentsState,
+        to: ChannelContentsState,
+        order: sync::atomic::Ordering,
+    ) -> ChannelState {
+        let prev: ChannelState = if to as u32 > from as u32 {
+            let inc = to as u32 - from as u32;
+            self.0.fetch_add(inc, order).into()
+        } else {
+            let dec = from as u32 - to as u32;
+            self.0.fetch_sub(dec, order).into()
+        };
+        assert_eq!(prev.contents_state, from);
+        prev
+    }
+
+    fn set_sleeper(&self, order: sync::atomic::Ordering) -> ChannelState {
+        let prev = ChannelState::from(self.0.fetch_add(HAS_SLEEPER, order));
+        assert!(!prev.has_sleeper);
+        prev
+    }
+
+    fn clear_sleeper(&self, order: sync::atomic::Ordering) -> ChannelState {
+        let prev = ChannelState::from(self.0.fetch_sub(HAS_SLEEPER, order));
+        assert!(prev.has_sleeper);
+        prev
+    }
+
+    fn set_writer_dead(&self, order: sync::atomic::Ordering) -> ChannelState {
+        let prev = ChannelState::from(self.0.fetch_add(WRITER_DIED, order));
+        assert!(!prev.writer_died);
+        prev
+    }
```

Comment:
Document the requirements of these functions? For example `set_sleeper` requires that the sleeper bit is unset, etc.